### ISSUE TITLE
crawling doesn't work

### ISSUE THREAD

**Issue by jenggoy** [francccisss/zensearch#2]
can you help me when i put link in the crawler i got error
![crawler_error](https://github.com/user-attachments/assets/d7946c30-3868-497b-896c-d5b644ef35ab)
i already restart the application but it won't help

**Comment by francccisss** [comment]
are you using docker compose? and if so do you mind sending me a screenshot of the terminal where docker compose is running when you try to crawl

**Comment by jenggoy** [comment]
yes i'm using docker compose. the terminal didn't show any log when i try to crawl 
![image](https://github.com/user-attachments/assets/de751307-dbb8-4827-94a6-caa1c876806d)
but when i try to search something it send logs
![image](https://github.com/user-attachments/assets/e9b23429-308f-438e-b7c9-3eeedab174e2)

**Comment by francccisss** [comment]
Seems like that crawler was not initialized for some reason, it could be that this is an older version, some time ago I forgot to include the crawler service in the docker-compose.yaml file lol, you could send a screenshot to me if the crawler is included in the docker-compose.yaml file in the root directory, if it is commented then its an old version so just clone the recent one in `main` branch

**Comment by francccisss** [comment]
![image](https://github.com/user-attachments/assets/27a79e8b-3fa9-44e0-8cf8-7de022c4ffe0)

**Comment by jenggoy** [comment]
but the crawler is already there. here's the docker-compose.yaml :
services:
  rabbitmq:
    image: rabbitmq
    healthcheck: &hck
      test: ["CMD","echo","Testing"]
      interval: 10s
      retries: 5
      start_period: 30s
      timeout: 10s
    networks:
      - zennet

  selenium:
    image: selenium/standalone-chrome
    shm_size: 2g
    healthcheck:
      <<: *hck
    networks:
      - zennet

  web:
    build:
      context: ./express-server/
      dockerfile: Dockerfile
    depends_on:
      rabbitmq:
        condition: service_healthy
    ports:
      - "8080:8080"
    restart: always
    networks:
      - zennet

  db:
    build:
      context: ./database/
      dockerfile: Dockerfile
    depends_on:
      rabbitmq:
        condition: service_healthy
    restart: always
    networks:
      - zennet
    volumes:
      - db_volume:/app/data/

  searchengine:
    build:
      context: ./search-engine/
      dockerfile: Dockerfile
    depends_on:
      rabbitmq:
        condition: service_healthy
    restart: always
    networks:
      - zennet

  crawler:
    build:
      context: ./crawler/
      dockerfile: Dockerfile
    depends_on:
      rabbitmq:
        condition: service_healthy
      selenium:
        condition: service_healthy
    restart: always
    networks:
        - zennet

networks:
  zennet:
    driver: bridge
volumes:
  db_volume:

**Comment by jenggoy** [comment]
but the crawler is alreaady included in the docker-compose.yaml 
![image](https://github.com/user-attachments/assets/d6ce7a7f-3aa6-4dc0-bb0d-d4fce41efe06)

**Comment by francccisss** [comment]
Hmm that's weird could you: 
- run zensearch via docker compose
- wait for rabbitmq to finish its start up operation
- look for `crawler-1` logs where it says "Successfully connected to rabbitmq" or any logs from crawler on startup

**Comment by jenggoy** [comment]
here you go
![image](https://github.com/user-attachments/assets/6230d593-19aa-4243-af82-2b2ae09bfa2b)
and this is the last log from crawler-1
![image](https://github.com/user-attachments/assets/b25c3431-db66-46c5-a37e-9c156198a08c)
it says "successfully connected to rabbitmq" and i try to crawl again but i still got the error message

**Comment by jenggoy** [comment]
here you go
![image](https://github.com/user-attachments/assets/4d5e9c5a-b758-43b1-b888-2ad1acc6ad9d)
this is the last log from crawler-1
![image](https://github.com/user-attachments/assets/cac5ef8e-88a5-44da-af37-46142ac0c25b)
it says "Successfully connected to rabbitmq" and i try to crawl again but i stll got the error message

**Comment by francccisss** [comment]
Thanks for the logs i'll look into it

**Comment by francccisss** [comment]
I got the same error after I removed all the images and containers for zensearch and called `docker compose up`, and got so many errors from other services saying they're not connected to the rabbitmq, but after re-running it using: 

`docker compose down`
`docker compose up`

The error is gone and I can use it normally and the crawler is connected correctly. It seems that other services are not waiting for the rabbitmq to finish it's startup operation on the first try using `docker compose up`, so these services `crawler`, `search-engine` etc... are interleaving with the rabbitmq startup.

This could be an error from `docker-compose.yaml` configuration file, so for now try running the commands above if it helps.

#### Services starting before rabbitmq:
##### both search engine and crawler started too early
![Screenshot from 2024-12-03 19-12-50](https://github.com/user-attachments/assets/89d0043e-13da-46cf-b047-55a3fa443899)
![image](https://github.com/user-attachments/assets/e4c6472b-1b6f-4c71-949e-e34a498d435f)

##### Rabbitmq finished later
The "Time to start" line is the end of the startup operations of rabbitmq
![Screenshot from 2024-12-03 19-11-25](https://github.com/user-attachments/assets/3a6eddc1-00aa-4009-9dd3-eebdaa83f983)


This error can be recreated by removing images and containers and the `zensearch_zennet` bridge network, and call `docker compose up` again.

**Comment by jenggoy** [comment]
hi i have followed what you said but i still can't crawl the url. i still got the error message
![image](https://github.com/user-attachments/assets/c4c94483-4a2d-4353-895b-ae1c25af377e)
the logs didn't show anything neither

**Comment by francccisss** [comment]
Hello, I've managed to fix the problem, it was because of the misuse of `aliases` and `merging` syntax in the docker compose config, that's why other services were not waiting for the rabbitmq to finish first before they start can you try it out now?

**Comment by francccisss** [comment]
Hi I've managed to fix it by implementing each service their own retry polling mechanism to wait for the message broker to finish its start up operation, if you have time please pull the updated commit and test it out if it works for you, and if so I will mark this as closed thanks!

**Comment by jenggoy** [comment]
Hello, i still can't crawl the url. the error message still same as before it asked me to restart the application and i already check all containers while crawling, here's the log for each container
crawler log:
![WhatsApp Image 2024-12-09 at 13 12 28_fdf0385b](https://github.com/user-attachments/assets/de9cda0b-7385-4b61-be60-65b2b3b01cb2)
db logs:
![WhatsApp Image 2024-12-09 at 13 13 54_38df450f](https://github.com/user-attachments/assets/258552ed-ac85-42b8-bcf9-f836b418283a)
rabbitmq logs:
![WhatsApp Image 2024-12-09 at 13 16 09_96b591f6](https://github.com/user-attachments/assets/965ee476-a7e8-4a4e-8df2-e7056278a818)
searchengine logs:
![WhatsApp Image 2024-12-09 at 13 24 57_e2ab51c7](https://github.com/user-attachments/assets/5dc7a7fe-073b-44b0-a767-9e33118a05fa)
selenium logs:
![WhatsApp Image 2024-12-09 at 13 25 21_9d6bf197](https://github.com/user-attachments/assets/beb4ed09-b097-4b91-8271-623cbcab0a9f)
web logs:
![image](https://github.com/user-attachments/assets/9ca9f8a6-f306-48fc-a288-4de4d7c6577e)
I have tried crawling the URL while checking each container, and none of them displayed any logs at all.
